src/TLuaInterpreter.cpp

```diff
@@ -4193,27 +4193,71 @@ int TLuaInterpreter::connectToServer(lua_State* L)
 // Documentation: https://wiki.mudlet.org/w/Manual:Lua_Functions#setRoomIDbyHash
 int TLuaInterpreter::setRoomIDbyHash(lua_State* L)
 {
-    int id;
     if (!lua_isnumber(L, 1)) {
         lua_pushfstring(L, "setRoomIDbyHash: bad argument #1 type (room id as number expected, got %s!)", luaL_typename(L, 1));
         return lua_error(L);
     }
-    id = lua_tonumber(L, 1);
+    qint64 id = lua_tointeger(L, 1);
     if (!lua_isstring(L, 2)) {
-        lua_pushfstring(L, "setRoomIDbyHash: bad argument #2 type (hash as string expected, got %s)", luaL_typename(L, 2));
+        lua_pushfstring(L, "setRoomIDbyHash: bad argument #2 type (hash as string expected, got %s!)", luaL_typename(L, 2));
         return lua_error(L);
     }
     QString hash = QString::fromUtf8(lua_tostring(L, 2));
+    // This reflects prior code that would silently replace an existing hash
+    // with a new roomId
+    bool allowCollisions = true;
+    if (lua_gettop(L) > 2 ) {
+        if (!lua_isboolean(L, 3)) {
+            lua_pushfstring(L, "setRoomIDbyHash: bad argument #3 type (permit collisions as boolean is optional {defaults to true if omitted}, got %s!)", luaL_typename(L, 3));
+            return lua_error(L);
+        }
+        allowCollisions = lua_toboolean(L, 3);
+    }
+
     Host& host = getHostFromLua(L);
-    if (host.mpMap->mpRoomDB->roomIDToHash.contains(id)) {
-        host.mpMap->mpRoomDB->hashToRoomID.remove(host.mpMap->mpRoomDB->roomIDToHash[id]);
+    auto pR = host.mpMap->mpRoomDB->getRoom(static_cast<int>(id));
+    if (!pR) {
+        lua_pushnil(L);
+        lua_pushfstring(L, "room id %d does not exist", id);
+        return 2;
     }
-    if (host.mpMap->mpRoomDB->hashToRoomID.contains(hash)) {
-        host.mpMap->mpRoomDB->roomIDToHash.remove(host.mpMap->mpRoomDB->hashToRoomID[hash]);
+
+    if (!pR->mHash.isEmpty()) {
+        // Clean old hash key from table - because this is a QMultiHash we need
+        // to specify BOTH key and value to only remove the value for the one
+        // id - during development code runs check that we have actually done
+        // something -  and moan if we haven't:
+        if (!host.mpMap->mpRoomDB->hashToRoomID.remove(pR->mHash, id)) {
```

Comment:
I am not sure why you think that all the roomIds which the hash is associated with **should** be removed?

It removes the hash for the specified room id - whether because a *different* hash is going to be used for *that* roomId or because the hash for that roomId is to be cleared for whatever purpose...

Reply:
Not all, just the ones that match the id we're removing. See comment below.